File: java/src/main/java/concepts/SinglyListNode.java
package concepts;

public final class SinglyListNode extends ListNode<SinglyListNode> {
    public static final SinglyListNode ZERO = new SinglyListNode(0);

    public SinglyListNode() {
        super(0);
    }

    public SinglyListNode(int value) {
        super(value);
    }

    public SinglyListNode(int value, SinglyListNode next) {
        super(value);
        this.next = next;
    }

    public static SinglyListNode of(int... vals) {
        if (vals == null || vals.length == 0) {
            return null;
        }
        SinglyListNode head = new SinglyListNode(vals[0]);
        SinglyListNode next = head;
        for (int i = 1; i < vals.length; i++) {
            next.next = new SinglyListNode(vals[i]);
            next = next.next;
        }
        return head;
    }
}

File: kotlin/src/main/kotlin/concepts/SinglyListNode.kt
package concepts

class SinglyListNode : ListNode<SinglyListNode> {
    constructor() : super(0)

    constructor(`val`: Int) : super(`val`)

    constructor(`val`: Int, next: SinglyListNode?) : super(`val`) {
        this.next = next
    }

    companion object {
        val ZERO = SinglyListNode(0)

        fun of(vararg vals: Int): SinglyListNode? {
            if (vals.isEmpty()) {
                return null
            }
            val head = SinglyListNode(vals[0])
            var next = head
            for (i in 1 until vals.size) {
                next.next = SinglyListNode(vals[i])
                next = next.next!!
            }
            return head
        }
    }
}
